rust/parquet/src/record/reader.rs

```diff
@@ -15,687 +15,1124 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! Contains implementation of record assembly and converting Parquet types into
-//! [`Row`](`::record::api::Row`)s.
-
-use std::{collections::HashMap, fmt, rc::Rc};
+//! Contains implementation of record assembly and converting Parquet types into Rust
+//! types.
+//!
+//! Readers wrap one or more columns'
+//! [`TypedTripletIter`](super::triplet::TypedTripletIter)s, and map their values from
+//! physical type to logical type. They're also responsible for correctly accessing fields
+//! that are optional or repeated.
+
+use fxhash::FxBuildHasher;
+use linked_hash_map::LinkedHashMap;
+use std::{
+    collections::HashMap, convert::TryInto, error::Error, marker::PhantomData, mem,
+    sync::Arc,
+};
 
-use crate::basic::{LogicalType, Repetition};
+use super::{
+    triplet::TypedTripletIter,
+    types::{
+        Bson, Date, Enum, Group, Json, List, Map, Root, Time, Timestamp, Value,
+        ValueRequired,
+    },
+    Predicate, Reader, Record,
+};
+use crate::column::reader::ColumnReader;
+use crate::data_type::{
+    BoolType, ByteArrayType, Decimal, DoubleType, FixedLenByteArrayType, FloatType,
+    Int32Type, Int64Type, Int96, Int96Type,
+};
 use crate::errors::{ParquetError, Result};
 use crate::file::reader::{FileReader, RowGroupReader};
-use crate::record::{
-    api::{make_list, make_map, make_row, Field, Row},
-    triplet::TripletIter,
-};
-use crate::schema::types::{ColumnPath, SchemaDescPtr, SchemaDescriptor, Type, TypePtr};
+use crate::schema::types::ColumnPath;
 
 /// Default batch size for a reader
 const DEFAULT_BATCH_SIZE: usize = 1024;
 
-/// Tree builder for `Reader` enum.
-/// Serves as a container of options for building a reader tree and a builder, and
-/// accessing a records iterator [`RowIter`].
-pub struct TreeBuilder {
-    // Batch size (>= 1) for triplet iterators
-    batch_size: usize,
+// ----------------------------------------------------------------------
+// Implementations for "anonymous" sum types
+
+impl<A, B> Reader for sum::Sum2<A, B>
```

Comment:
This is pretty confusing - why we implement reader on `Sum2` and `Sum3`?

Reply:
If you're familiar with bluss's [`Either`](https://github.com/bluss/either) crate, then `Sum2` is just the same – an enum with two generic variants. `Sum3` is the same, but with three generic variants. Rather than creating a new enum every time you need "an enum that can contain either of two/three things", they let you reuse the generic `Either<A, B>`/`Sum2<A, B>`/`Sum3<A, B, C>`.

`src/record/types/time.rs` requires "an enum that can contain either of two things" and "an enum that can contain either of three things". Up until a couple of days ago when I implemented a temporary workaround to a bug in `cargo doc`, `src/record/types/decimal.rs` also required "an enum that can contain either of three things". Rather than create three new specific enums, I chose to reuse the generic enums `Sum2` and `Sum3` instead.

I hesitated over this as there are only three enums required (and two right now until the `cargo doc` bug is resolved) so there's minimal gain. If however the codebase grows and more such enums are required, it makes more sense.

Let me know what you think. I'm going to go ahead and add in my rationale where you've commented, but I'm more than happy to replace them with seperate specific enums.